### PR TITLE
Fixing Resistor Models to work properly for ngspice

### PR DESCRIPTION
**Pull request by atorkmabrains**
Fixing Resistor Models to work properly for ngspice

### CODE REVIEW

File: models/ngspice/testing/regression/.spiceinit [hunk 1]
```diff
@@ -1,2 +1,3 @@
 ************
 set ngbehavior=hs
+set numthreads=8
```
Comment (mithro) [inline models/ngspice/testing/regression/.spiceinit:3]: Changing the number of threads is unrelated to the resistor change?

File: models/ngspice/testing/regression/resistor_r/device_netlists/2term_res_a.spice [hunk 2]
```diff
@@ -1,44 +0,0 @@
-**************************
```
Comment (mithro) [inline models/ngspice/testing/regression/resistor_r/device_netlists/2term_res_a.spice:1]: Why are the resistor regression tests being removed?

**Review by mithro — CHANGES_REQUESTED** [review]
Some small fixes needed.

File: models/ngspice/testing/regression/.spiceinit (hunk 1 above)
Comment (atorkmabrains) [reply models/ngspice/testing/regression/.spiceinit:3]: It helps speed up the testing.

File: models/ngspice/testing/regression/resistor_r/device_netlists/2term_res_a.spice (hunk 2 above)
Comment (atorkmabrains) [reply models/ngspice/testing/regression/resistor_r/device_netlists/2term_res_a.spice:1]: They are not removed. They are moved and changed.

**Comment by atorkmabrains** [comment]
@mithro replied.

File: models/ngspice/testing/regression/.spiceinit (hunk 1 above)
Comment (mithro) [reply models/ngspice/testing/regression/.spiceinit:3]: I'm not claiming it doesn't speed up testing.

I'm saying a change which speeds up testing should not be mixed into a change which is suppose to fix resistors.

Please split the changes into one pull request per feature change

File: models/ngspice/testing/regression/resistor_r/device_netlists/2term_res_a.spice (hunk 2 above)
Comment (proppy) [reply models/ngspice/testing/regression/resistor_r/device_netlists/2term_res_a.spice:1]: what's the new test equivalent to this one?

File: models/ngspice/testing/regression/resistor_r/device_netlists/3term_res_a.spice [hunk 3]
```diff
@@ -1,44 +0,0 @@
-**************************
```
Comment (proppy) [inline models/ngspice/testing/regression/resistor_r/device_netlists/3term_res_a.spice:1]: what's the new test equivalent to this one?

File: models/ngspice/testing/regression/resistor_r/device_netlists/3term_res_b.spice [hunk 4]
```diff
@@ -1,44 +0,0 @@
-**************************
```
Comment (proppy) [inline models/ngspice/testing/regression/resistor_r/device_netlists/3term_res_b.spice:1]: what's the new test equivalent to this one?

File: models/ngspice/testing/regression/resistor_r/device_netlists/res_op_analysis.spice [hunk 5]
```diff
@@ -15,30 +15,23 @@
...
+XR1 top {{terminals}} {{device}} r_width={{width}}u r_length={{length}}u m=1
 
 .temp {{temp}}
 .options tnom={{temp}}
-      
-r1 n p 1M
-xrn p 0 {{device}} r_width={{width}}u r_length={{length}}u
-
-
-.print dc Rs=par('(1M*V(p)/(V(n)-V(p)))/{{length}}u*{{width}}u')
```
Comment (proppy) [inline models/ngspice/testing/regression/resistor_r/device_netlists/res_op_analysis.spice:27]: is the print statement not needed anymore?

File: models/ngspice/testing/regression/resistor_r/device_netlists/res_op_analysis.spice [hunk 6]
```diff
@@ -15,30 +15,23 @@
 * See the License for the specific language governing permissions and
 * limitations under the License.
 
-Vn n 0 DC {{sign}}3.3
+Vin top GND {{voltage}}
+XR1 top {{terminals}} {{device}} r_width={{width}}u r_length={{length}}u m=1
```
Comment (proppy) [inline models/ngspice/testing/regression/resistor_r/device_netlists/res_op_analysis.spice:19]: is this equivalent to the two `r1` and `xrn` statement removed below?

File: models/ngspice/testing/regression/resistor_r/models_regression.py [hunk 7]
```diff
@@ -20,295 +20,233 @@
...
+    process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
+    return float(process.communicate()[0][:-1].decode("utf-8").split(" ")[2])
+
 def call_simulator(file_name):
     """Call simulation commands to perform simulation.
     Args:
         file_name (str): Netlist file name.
     """
-    os.system(f"ngspice -b -a {file_name} -o {file_name}.log > {file_name}.log")
+    return os.system(f"ngspice -b -a {file_name} -o {file_name}.log > {file_name}.log")
```
Comment (proppy) [inline models/ngspice/testing/regression/resistor_r/models_regression.py:57]: shouldn't `Popen` be used instead?

File: models/ngspice/testing/regression/resistor_r/models_regression.py [hunk 8]
```diff
@@ -20,295 +20,233 @@
...
-    # Get dimensions used for each device 
-    dirpath = f"{device}_{r_sim}_{corner}_temp"
-    dimensions = pd.read_csv(f"{dirpath}/{device}.csv",usecols=["Temperature (C)" , "l (um)"])
-    if "rm" in device or "tm" in device:        
-        loops = dimensions["l (um)"].count()
-    else:
-        loops = 11
+def ext_const_temp_corners(dev_data_path, device, corners):
+    # Read Data
+    df = pd.read_excel(dev_data_path)
```
Comment (proppy) [inline models/ngspice/testing/regression/resistor_r/models_regression.py:61]: did those change from `csv` to `excel` file?

File: models/ngspice/testing/regression/resistor_r/models_regression.py [hunk 9]
```diff
@@ -20,295 +20,233 @@
...
-    df_final.to_csv (f"{dirpath}/Final_report_{r_sim}.csv", index = False)
-    out_report = pd.read_csv (f"{dirpath}/Final_report_{r_sim}.csv")
-    print ("\n",f"Max. mean error = {out_report['Mean error%'].max()}%")    
-    print ("=====================================================================================================================================================")
+        for future in concurrent.futures.as_completed(futures_list):
+            try:
+                data = future.result()
+                results.append(data)
+            except Exception as exc:
+                print('Test case generated an exception: %s' % (exc))
```
Comment (proppy) [inline models/ngspice/testing/regression/resistor_r/models_regression.py:162]: shouldn't this actually rethrow the exception so that the test fails?

**Review by proppy — COMMENTED** [review]
This seems to do a lot more than just fixing the resistance values (it alter significantly data processing and templating logic), I'd suggest documenting those change in the PR description, or splitting them in different changes if they are not related.

**Comment by atorkmabrains** [comment]
@proppy They are related. I changed the circuit template used for measuring the resistance. Nothing else.

**Pull request closed without merging**